### PR TITLE
Add style to pagination component

### PR DESCRIPTION
**Pull request by CharlyMartin**
This PR adds styling, ellipsis, and accessibility concerns to the pagination component.

### CODE REVIEW

**Comment by mirhamasala** [comment]
Super cool @CharlyMartin! ☀️ As early review comments go:

- Let's make this keyboard accessible
- Let's work on matching the styles closer to the Figma design (Vertical lines are a different color, chevrons seems a little thicker, and there's not so much whitespace around the numbers.)

Screenshot 2024-04-12 at 07 23 37

Screenshot 2024-04-12 at 07 22 56

File: src/app/_components/Pagination.tsx [hunk 1]
```diff
@@ -13,7 +17,13 @@ export function Pagination({
   currentPage,
   setCurrentPage,
 }: PaginationProps) {
-  const pageNumbers = Array.from({ length: pageCount }, (_, index) => index + 1)
+  const isXsDevice = useMediaQuery('only screen and (max-width : 640px)')
+
+  const collapsedRange = useCollapseRange(
+    pageCount,
+    currentPage,
+    isXsDevice ? 2 : 6,
```
Comment (mirhamasala) [inline src/app/_components/Pagination.tsx:25]: The `2` and `6` are like[ magic numbers](https://www.inf.unibz.it/~calvanese/teaching/05-06-ip/lecture-notes/uni04/node17.html#:~:text=A%20magic%20number%20is%20a,difficult%20to%20maintain%20and%20update.) - We can define them with variables.

File: src/app/_components/Pagination.tsx [hunk 2]
```diff
@@ -28,37 +38,63 @@ export function Pagination({
...
 
-      <button
-        className="rounded px-3 py-1 transition hover:bg-brand-400/30"
-        onClick={handleNext}
-      >
-        Next
-      </button>
+      <ul className="flex shrink grow justify-center gap-2 sm:gap-1">
+        {collapsedRange.map((item, index) => {
+          return (
```
Comment (mirhamasala) [inline src/app/_components/Pagination.tsx:59]: We can use an implicit return here.

File: src/app/_components/ScreenSizeIndicator.tsx [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    <div className="fixed bottom-1 left-1 z-50 rounded bg-zinc-800 px-1 font-mono text-xs text-white dark:bg-zinc-100 dark:text-zinc-900">
+      <div className="block sm:hidden">xs</div>
+      <div className="hidden sm:block md:hidden">sm</div>
+      <div className="hidden md:block lg:hidden">md</div>
+      <div className="hidden lg:block xl:hidden">lg</div>
+      <div className="hidden xl:block 2xl:hidden">xl</div>
+      <div className="hidden 2xl:block">2xl</div>
+    </div>
+  )
+}
```
Comment (mirhamasala) [inline src/app/_components/ScreenSizeIndicator.tsx:14]: Should be a different PR.

File: src/app/_hooks/useCollapseRange.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    )
+
+    return pageNumbers.slice(
+      Math.min(firstOrBeforeCurrentIndex, pageNumbers.length - range),
+      Math.max(lastOrAfterCurrentIndex, range),
+    )
+  }
+
+  // Ellipsis at the start only
+  if (needsStartEllipsis && !needsEndEllipsis) {
```
Comment (mirhamasala) [inline src/app/_hooks/useCollapseRange.ts:55]: If we're writing comments, it's a good indication that the code could be clearer.

This condition, `needsStartEllipsis && !needsEndEllipsis`, can be defined in a variable.

File: src/app/_hooks/useCollapseRange.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
+import { useMemo } from 'react'
+
+// Inspiration: https://atlassian.design/components/pagination
```
Comment (mirhamasala) [inline src/app/_hooks/useCollapseRange.ts:3]:
```suggestion
```

File: src/app/_hooks/useCollapseRange.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+export function useCollapseRange(
+  pageCount: number,
+  currentPage: number,
+  range: number = 7,
+): Array<number | typeof ELLIPSIS> {
+  const pageNumbers = useMemo(() => {
+    return Array.from({ length: pageCount }, (_, i) => i + 1)
+  }, [pageCount])
+
+  // Boolean
```
Comment (mirhamasala) [inline src/app/_hooks/useCollapseRange.ts:18]:
```suggestion
```

File: src/app/_hooks/useCollapseRange.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  const needEllipsis = pageNumbers.length > range
+  const needsStartEllipsis = currentPage - DISTANCE_FROM_START > 0
+  const needsEndEllipsis = currentPage + DISTANCE_FROM_END < pageNumbers.length
+
+  // page numbers
```
Comment (mirhamasala) [inline src/app/_hooks/useCollapseRange.ts:23]:
```suggestion
```

File: src/app/_hooks/useCollapseRange.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  const firstPageNumber = pageNumbers[0]
+  const lastPageNumber = pageNumbers[pageNumbers.length - 1]
+
+  // Indexes of page numbers
```
Comment (mirhamasala) [inline src/app/_hooks/useCollapseRange.ts:27]:
```suggestion
```

File: src/app/_hooks/useCollapseRange.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  const firstPageNumberIndex = pageNumbers.indexOf(firstPageNumber)
+  const currentPageIndex = pageNumbers.indexOf(currentPage)
+
+  // Early returns
+  if (range <= 0) throw new Error('range must be greater than 0')
+  if (!needEllipsis) return pageNumbers
+  if (range === 1) return [currentPage]
+
+  // If range is less than 4, we don't show any ellipsis
```
Comment (mirhamasala) [inline src/app/_hooks/useCollapseRange.ts:36]: Instead of having comments, could we abstract this into a function with appropriate naming?

File: src/app/_hooks/useCollapseRange.ts [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+    return pageNumbers.slice(
+      Math.min(firstOrBeforeCurrentIndex, pageNumbers.length - range),
+      Math.max(lastOrAfterCurrentIndex, range),
+    )
+  }
+
+  // Ellipsis at the start only
+  if (needsStartEllipsis && !needsEndEllipsis) {
+    const remainingArraySlots = range - 2
```
Comment (mirhamasala) [inline src/app/_hooks/useCollapseRange.ts:56]: Magic numbers floating around `2` and `4`  ✨

File: src/app/_hooks/useCollapseRange.ts [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+// Inspiration: https://atlassian.design/components/pagination
+
+const DISTANCE_FROM_START = 3
+const DISTANCE_FROM_END = 2
+const ELLIPSIS = '...'
+
+export function useCollapseRange(
+  pageCount: number,
+  currentPage: number,
+  range: number = 7,
```
Comment (mirhamasala) [inline src/app/_hooks/useCollapseRange.ts:12]: Magic number ✨ `7`

File: src/app/_components/Pagination.tsx [hunk 12]
```diff
@@ -28,37 +38,63 @@ export function Pagination({
...
-        className="rounded px-3 py-1 transition hover:bg-brand-400/30"
-        onClick={handlePrev}
-      >
-        Prev
-      </button>
+    <div className="flex w-full justify-evenly gap-4 rounded-lg bg-brand-300 p-1.5 text-brand-700">
+      <div className="flex">
+        <button
+          className="flex items-center rounded bg-brand-300 p-1 px-2 transition hover:bg-brand-400 focus-visible:outline focus-visible:outline-2 focus-visible:outline-white"
+          aria-label="Previous"
```
Comment (mirhamasala) [inline src/app/_components/Pagination.tsx:45]: The `aria-label` could be more descriptive `Go to previous page`

File: src/app/_components/Pagination.tsx (hunk 1 above)
Comment (CharlyMartin) [reply src/app/_components/Pagination.tsx]: Love it! 

I added a `useResponsiveRange` to compute the range's size based on the screen size. Hopefully, the numbers make more sense in that context.

File: src/app/_components/ScreenSizeIndicator.tsx (hunk 3 above)
Comment (CharlyMartin) [reply src/app/_components/ScreenSizeIndicator.tsx]: Yes, you're right. I reverted it.

File: src/app/_components/Pagination.tsx (hunk 2 above)
Comment (CharlyMartin) [reply src/app/_components/Pagination.tsx]: ✅

File: src/app/_hooks/useCollapseRange.ts (hunk 4 above)
Comment (CharlyMartin) [reply src/app/_hooks/useCollapseRange.ts:42]: > If we're writing comments, it's a good indication that the code could be clearer.

I see what you mean, but for long/complex code, I think comments add a visual hierarchy to the file so that it's more accessible for the reader, in the same way headings structure paragraphs.

Thoughs?

File: src/app/_hooks/useCollapseRange.ts (hunk 10 above)
Comment (CharlyMartin) [reply src/app/_hooks/useCollapseRange.ts:43]: What do you mean?

File: src/app/_hooks/useCollapseRange.ts (hunk 4 above)
Comment (mirhamasala) [reply src/app/_hooks/useCollapseRange.ts:42]: I don't agree. If the code is long and complex, then it should be broken up in shorter and simpler code. Abstract bits into functions, etc. I recommend reading chapter 4 from Clean Code by Robert C. Martin. (I recommend the entire book - until about chapter 8 - After that, it's very Java specific if I remember correctly.)

>So when you find yourself in a position where you need to write a comment, think it through and see whether there isn’t some way to turn the tables and express yourself in code. Every time you express yourself in code, you should pat yourself on the back. Every time you write a comment, you should grimace and feel the failure of your ability of expression.

File: src/app/_hooks/useCollapseRange.ts (hunk 5 above)
Comment (mirhamasala) [reply src/app/_hooks/useCollapseRange.ts:3]: Can be added to the PR message if needed.

**Comment by CharlyMartin** [comment]
Thank you for the review @mirhamasala ✨ 

I've implemented most of your feedback with these 3 commits:
1. https://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/54/commits/317c8538d7838d98f8144ed793b85caeb1188b4c
2. https://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/54/commits/829ad66d606e4b0041e0d0ebf5f2bfb65ce8bdae
3. https://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/54/commits/aea1d6a0dffad1ce0857a8dddc249481b3e79af8

I also removed the `ScreenSizeIndicator` with commit https://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/54/commits/bb8ba801fd73b1e9fa073bf6ddcd4d14ba8a5f45, we'll add it later via a separate [ticket](https://www.notion.so/filecoin/FF-V4-Add-ScreenSizeIndicator-component-e85bd71442af4670a40d52ac4cae86b5?pvs=4).

Regarding the keyboard accessibility of the page, something is off indeed so we can explore a fix in a separate [ticket](https://www.notion.so/filecoin/FF-V4-Explore-keyboard-accessibility-of-blog-page-and-pagination-component-981a71bbad4d4a8d91c8ee7868dd754b?pvs=4) as well

Finally, I added a few more comments/questions.

File: src/app/_components/Pagination.tsx [hunk 13]
```diff
@@ -13,52 +17,110 @@ export function Pagination({
...
-      >
-        Next
-      </button>
-    </div>
+      <div className="flex">
+        <div className="flex items-center">
+          <LineVertical
+            size={24}
+            weight="light"
+            className="text-[#08072E]/50"
```
Comment (mirhamasala) [inline src/app/_components/Pagination.tsx:104]:
```suggestion
            className="text-brand-800/50"
```

File: src/app/_components/Pagination.tsx [hunk 14]
```diff
@@ -13,52 +17,110 @@ export function Pagination({
...
-          </button>
+            {typeof item === 'number' ? (
+              <button
+                className={clsx(
+                  'h-full w-full rounded focus-visible:outline focus-visible:outline-2 focus-visible:outline-white',
+                  item === currentPage
+                    ? 'bg-brand-800 text-brand-100'
+                    : 'bg-brand-300 text-brand-700 hover:bg-brand-400',
+                )}
+                aria-label={'Page ' + item}
```
Comment (mirhamasala) [inline src/app/_components/Pagination.tsx:85]:
```suggestion
                aria-label={`Go to page ${item}`}
```

File: src/app/_components/Pagination.tsx [hunk 15]
```diff
@@ -13,52 +17,110 @@ export function Pagination({
...
-              page === currentPage
-                ? 'bg-brand-800 text-brand-100'
-                : 'bg-brand-300 text-brand-700',
-            )}
-            onClick={() => handlePageChange(page)}
+        <div className="flex items-center">
+          <LineVertical
+            size={24}
+            weight="light"
+            className="text-[#08072E]/50"
```
Comment (mirhamasala) [inline src/app/_components/Pagination.tsx:65]:
```suggestion
            className="text-brand-800/50"
```

**Comment by CharlyMartin** [comment]
@mirhamasala I've reworked the `useCollapseRange` hook, now called `useVisiblePages`, to be more accessible and easier to read.

Let me know what you think :)

**Comment by mirhamasala** [comment]
> Let me know what you think :)

@charlymartin - I think it's amazing. Thank you for making it happen. ✨
